certora/specs/MintBurnEthereum.spec

```diff
@@ -0,0 +1,135 @@
+*/
+rule onlyAuthorizedCanTransfer(env e, method f) filtered {
+    f-> f.selector != sig:upgradeToAndCall(address, bytes).selector
+} {
+    requireInvariant totalSupplyIsSumOfBalances();
+    requireInvariant balancesLTEqTotalSupply();
+    requireInvariant twoBalancesLTEqTotalSupply();
+
+    calldataarg args;
+    address account;
+
+    uint256 allowanceBefore = allowance(account, e.msg.sender);
+    uint256 balanceBefore   = balanceOf(account);
+    f(e, args);
+    uint256 balanceAfter    = balanceOf(account);
+
+    assert (
+        balanceAfter < balanceBefore
+    ) => (
+        f.selector == sig:burn(uint256).selector ||
+        e.msg.sender == account ||
+        balanceBefore - balanceAfter <= to_mathint(allowanceBefore)
+    );
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: only mint and burn can change total supply                                                                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noChangeTotalSupply(env e, method f) filtered {
+    f-> f.selector != sig:upgradeToAndCall(address, bytes).selector
+} {
+    requireInvariant totalSupplyIsSumOfBalances();
+    requireInvariant balancesLTEqTotalSupply();
+
+    calldataarg args;
+
+    uint256 totalSupplyBefore = totalSupply();
+    f(e, args);
+    uint256 totalSupplyAfter = totalSupply();
+
+    assert totalSupplyAfter > totalSupplyBefore => f.selector == sig:mint(address,uint256).selector || f.selector == sig:initialize(address, address).selector;
+    assert totalSupplyAfter < totalSupplyBefore => f.selector == sig:burn(uint256).selector;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: mint behavior and side effects                                                                               │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule mint(env e) {
+    requireInvariant totalSupplyIsSumOfBalances();
+    requireInvariant balancesLTEqTotalSupply();
+    assert isTotalSupplyGTEqSumOfVotingPower();
+    requireInvariant zeroAddressNoVotingPower();
+    require nonpayable(e);
+
+    address to;
+    address other;
+    uint256 amount;
+
+    // cache state
+    uint256 toBalanceBefore    = balanceOf(to);
+    uint256 toVotingPowerBefore = delegatedVotingPower(delegatee(to));
+    uint256 otherBalanceBefore = balanceOf(other);
+    uint256 totalSupplyBefore  = totalSupply();
+
+    // Safe require thats avoid absurd counter-examples.
+    require toVotingPowerBefore <= sumOfVotingPower;
+
+    // run transaction
+    mint@withrevert(e, to, amount);
+
+    // check outcome
+    if (lastReverted) {
```

Comment:
I think we settled on trying to verify all the revert conditions, ie verify something of the form `lastReverted <=>`, as you have done in #98. Let's do it here too

Reply:
For the record, the revert conditions are checked in #98 to preserve simpler diffs w.r.t to OZ's original spec.

Reply:
let's not change anything now, but I think that there is low value is copying OZ's spec (vs having a clean spec)